pallets/summary/src/lib.rs

```diff
@@ -0,0 +1,1438 @@
+                        *root_data.tx_id.as_ref().expect("Non empty roots have valid hash"),
+                        root_data.added_by.ok_or(Error::<T>::CurrentSlotValidatorNotFound)?,
+                        voting_session.state()?.confirmations,
+                    );
+
+                if let Err(result) = result {
+                    log::error!("❌ Error Submitting Tx: {:?}", result);
+                    Err(result)?
+                }
+                // There are a couple possible reasons for failure.
+                // 1. We fail before sending to T1: likely a bug on our part
+                // 2. Quorum mismatch. There is no guarantee that between accepting a root and
+                // submitting it to T1, the tier2 session hasn't changed and with it
+                // the quorum, making ethereum-transactions reject it
+                // In either case, we should not slash anyone.
+            }
+            // If we get here, then we did not get an error when submitting to T1.
+
+            create_and_report_summary_offence::<T>(
+                &reporter,
+                &vote.nays,
+                SummaryOffenceType::RejectedValidRoot,
+            );
+
+            let next_block_to_process =
+                safe_add_block_numbers::<T::BlockNumber>(root_id.range.to_block, 1u32.into())
+                    .map_err(|_| Error::<T>::Overflow)?;
+
+            <NextBlockToProcess<T>>::put(next_block_to_process);
+            <Roots<T>>::mutate(root_id.range, root_id.ingress_counter, |root| {
+                root.is_validated = true
+            });
+            <SlotOfLastPublishedSummary<T>>::put(Self::current_slot());
+
+            Self::deposit_event(Event::<T>::SummaryRootValidated {
+                root_hash: root_data.root_hash,
+                ingress_counter: root_id.ingress_counter,
+                block_range: root_id.range,
+            });
+        } else {
+            // We didn't get enough votes to approve this root
+
+            let root_creator =
+                root_data.added_by.ok_or(Error::<T>::CurrentSlotValidatorNotFound)?;
+            create_and_report_summary_offence::<T>(
+                &reporter,
+                &vec![root_creator],
+                SummaryOffenceType::CreatedInvalidRoot,
+            );
+
+            create_and_report_summary_offence::<T>(
+                &reporter,
+                &vote.ayes,
+                SummaryOffenceType::ApprovedInvalidRoot,
+            );
+        }
+
+        <PendingApproval<T>>::remove(root_id.range);
+
+        // When we get here, the root's voting session has ended and it has been removed from
+        // PendingApproval If the root was approved, it is now marked as validated.
+        // Otherwise, it stays false If there was an error when submitting to T1, none of
+        // this happened and it is still pending and not validated In either case, the whole
+        // voting history remains in storage
```

Comment:
```suggestion
        // When we get here, the root's voting session has ended and it has been removed from
        // PendingApproval. If the root was approved, it is now marked as validated.
        // Otherwise, it stays false
        // If there was an error when submitting to T1, none of
        // this happened and it is still pending and not validated.
        // In either case, the whole voting history remains in storage
```